src/mscorlib/shared/System/Text/StringBuilder.cs

```diff
@@ -790,6 +790,123 @@ public StringBuilder Append(string value, int startIndex, int count)
+            }
+            return this;
+        }
+
+        private void AppendCore(StringBuilder value, int startIndex, int count)
+        {
+            int newLength = Length + count;
+            if ((uint)newLength > (uint)m_MaxCapacity)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_SmallMaxCapacity);
+            }
+
+            while (count > 0)
+            {
+                int length = Math.Min(m_ChunkChars.Length - m_ChunkLength, count);
+                if (length == 0)
+                {
+                    ExpandByABlock(count);
+                    length = Math.Min(m_ChunkChars.Length - m_ChunkLength, count);
+                }
+
+                StringBuilder chunk = value;
+                int sourceEndIndex = startIndex + length;
+                int curDestIndex = length;
+                unsafe
+                {
+                    fixed (char* destinationPtr = &m_ChunkChars[m_ChunkLength])
+                    {
+                        while (curDestIndex > 0)
+                        {
+                            int chunkEndIndex = sourceEndIndex - chunk.m_ChunkOffset;
+                            if (chunkEndIndex >= 0)
+                            {
+                                if (chunkEndIndex > chunk.m_ChunkLength)
+                                    chunkEndIndex = chunk.m_ChunkLength;
+
+                                int countLeft = curDestIndex;
+                                int chunkCount = countLeft;
+                                int chunkStartIndex = chunkEndIndex - countLeft;
+                                if (chunkStartIndex < 0)
+                                {
+                                    chunkCount += chunkStartIndex;
+                                    chunkStartIndex = 0;
+                                }
+                                curDestIndex -= chunkCount;
+
+                                if (chunkCount > 0)
+                                {
+                                    // work off of local variables so that they are stable even in the presence of race conditions
+                                    char[] sourceArray = chunk.m_ChunkChars;
+
+                                    // Check that we will not overrun our boundaries. 
+                                    if ((uint)(chunkCount + curDestIndex) <= (uint)length && (uint)(chunkCount + chunkStartIndex) <= (uint)sourceArray.Length)
+                                    {
+                                        fixed (char* sourcePtr = &sourceArray[chunkStartIndex])
+                                            string.wstrcpy(destinationPtr + curDestIndex, sourcePtr, chunkCount);
+                                    }
+                                    else
+                                    {
+                                        throw new ArgumentOutOfRangeException(nameof(chunkCount));
```

Comment:
@danmosemsft  not able to find any appropriate helper function for this. Should I choose ThrowArgumentOutOfRangeException, ThrowArgumentOutOfRange_IndexException() ?

Reply:
Either is fine .. this will never be hit except in broken apps, I think.